### PR TITLE
consensus: add calculation for proposal step waits from pbts 

### PR DESCRIPTION
**Pull request by williambanfield**
This pull requests implements helper functions to for calculating sleep durations during the propose phase.

For more information on the inequalities in these functions, please see the proposer-based timestamps ADR. Specifically, this pull requests implements the [proposer waits](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-071-proposer-based-timestamps.md#proposer-waits) and [changes to propose step timeout](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-071-proposer-based-timestamps.md#changes-to-the-propose-step-timeout) calculations.

Related to: #6942

### CODE REVIEW

File: internal/consensus/state.go [hunk 1]
```diff
@@ -2412,3 +2412,31 @@ func repairWalFile(src, dst string) error {
 
 	return nil
 }
+
+// proposerWaitTime determines how long the proposer should wait to propose its next block.
+// Block times must be monotonically increasing, so if the block time of the previous
+// block is larger than the proposer's current time, then the proposer will sleep
+// until its local clock exceeds the previous block time.
```
Comment (creachadair) [inline internal/consensus/state.go:2419]: Optional formatting suggestion: Document what zero means, and separate explanation from the description of behaviour:

```suggestion
// proposerWaitTime determines how long the proposer should wait to propose its next block.
// If the result is zero, a block can be proposed immediately.
//
// Block times must be monotonically increasing, so if the block time of the previous
// block is larger than the proposer's current time, then the proposer will sleep
// until its local clock exceeds the previous block time.
```

File: internal/consensus/state.go [hunk 2]
```diff
@@ -2412,3 +2412,31 @@ func repairWalFile(src, dst string) error {
...
+	return h.Time.Sub(t)
+}
+
+// proposalStepWaitingTime determines how long a validator should wait for a block
+// to be sent from a proposer. This duration is determined as a result of the
+// previous block's time as well as by the Accuracy and MsgDelay timestamp parameters.
+// The validator's and the proposer's clocks should differ from eachother by at most
+// 2 * Accuracy and the proposal should take at most MsgDelay to arrive at the validator.
+// The validator must therefore wait at least 2 * Accuracy + MsgDelay for the proposal
+// to arrive.
```
Comment (creachadair) [inline internal/consensus/state.go:2434]:
```suggestion
// proposalStepWaitingTime determines how long a validator should wait for a block
// to be sent from a proposer. If zero, the validator should not wait.
//
// This duration is determined as a result of the previous block's time as well as
// by the Accuracy and MsgDelay timestamp parameters.
// The validator's and the proposer's clocks should differ from each other by at most
// 2 * Accuracy and the proposal should take at most MsgDelay to arrive at the validator.
// The validator must therefore wait at least 2 * Accuracy + MsgDelay for the proposal
// to arrive.
```

Can we say something about what happens if the "should" is violated? (I.e., in the event the clocks differ by more than that, or the actual delay is longer)

**Review by tychoish — COMMENTED** [review]
the new functions in `state.go` aren't actually called in any non-test functions, are you planning on doing a separate integration pass?

**Comment by williambanfield** [comment]
> the new functions in `state.go` aren't actually called in any non-test functions, are you planning on doing a separate integration pass?

yeah, I'd like to get all of the pieces in place first. Once things are integrated, they'll begin to impact consensus and I want to put everything in place and make sure it looks good/is correcet before I actually do that.

File: internal/consensus/state.go (hunk 2 above)
Comment (cmwaters) [reply internal/consensus/state.go:2434]: Yeah by:
> If zero, the validator should not wait.
You mean that the validator should prevote nil and advance to the next step

Comment (williambanfield) [reply internal/consensus/state.go]: I've reworded to hopefully more accurately reflect that goal of this function.

**Review by creachadair — APPROVED** [review]

**Review by tychoish — APPROVED** [review]

File: internal/consensus/state.go [hunk 3]
```diff
@@ -2412,3 +2412,38 @@ func repairWalFile(src, dst string) error {
...
+// To ensure that the validator waits long enough, it must wait until the previous
+// block's timestamp. It also must account for inaccuracy in its own clock, inaccuracy
+// in the proposer's clock and the amount of time for the message to be transmitted.
+//
+// The result of proposalStepWaitingTime is compared with the configured `timeout-propose` duration,
+// and the validator waits for whichever duration is larger before advancing to the next step
+// and prevoting nil.
+func proposalStepWaitingTime(lt tmtime.Source, h types.Header, tp types.TimestampParams) time.Duration {
+	t := lt.Now()
+	wt := h.Time.Add(2 * tp.Accuracy).Add(tp.MsgDelay)
```
Comment (cason) [inline internal/consensus/state.go:2444]: Why `Accuracy` here and not `Precision`?

**Review by cason — COMMENTED** [review]
I really did not understand the goal of the `proposalStepWaitingTime` method.

For me, the timeout propose event should be just scheduled to occur in `timeout_propose` duration plus `proposalStepWaitingTime`.

And if `timeout_propose < 2 * Precision + msgDelay`, well, we risk to prevote `nil` at a time at which we could still accept a  *timely* `Proposal`. This is a poor choice of parameters, not a wrong behavior.

File: internal/consensus/state.go [hunk 4]
```diff
@@ -2412,3 +2412,38 @@ func repairWalFile(src, dst string) error {
...
+
+// proposerWaitTime determines how long the proposer should wait to propose its next block.
+// If the result is zero, a block can be proposed immediately.
+//
+// Block times must be monotonically increasing, so if the block time of the previous
+// block is larger than the proposer's current time, then the proposer will sleep
+// until its local clock exceeds the previous block time.
+func proposerWaitTime(lt tmtime.Source, h types.Header) time.Duration {
+	t := lt.Now()
+	if t.After(h.Time) {
```
[the remark this answers is not in the document]
Comment (williambanfield) [reply internal/consensus/state.go]: Changed.

File: internal/consensus/state.go (hunk 3 above)
Comment (williambanfield) [reply internal/consensus/state.go]: Ah, this was from before our conversation. Changed to `Precision`

**Comment by williambanfield** [comment]
> I really did not understand the goal of the `proposalStepWaitingTime` method.
> 
> For me, the timeout propose event should be just scheduled to occur in `timeout_propose` duration plus `proposalStepWaitingTime`.
> And if `timeout_propose < 2 * Precision + msgDelay`, well, we risk to prevote `nil` at a time at which we could still accept a _timely_ `Proposal`. This is a poor choice of parameters, not a wrong behavior.

The `proposalStepWaitingTime` method calculates the `waitingTime` defined [here](https://github.com/tendermint/spec/blob/master/spec/consensus/proposer-based-timestamp/pbts-algorithm_001_draft.md#pbts-processing-step0) in the spec. I think we should discuss whether we want this to change to take use just the `timeout-propose` when `timeout-propose` is smaller than `waitingTime` on the upcoming spec changes. It will mean that picking the correct `timeout-propose` would affect liveness in a similar way it does now: if a block has not arrived by `timeout-propose` we prevote nil. This is the same as the current implementation, so I think that's a fine change, but I'd like further input from before making that change at the code level.

**Pull request merged**